File: app/src/main/java/ru/ushell/app/api/websocket/socketclient/TopicHandler.java
package ru.ushell.app.api.websocket.socketclient;

import java.util.HashSet;
import java.util.Set;

import ru.ushell.app.api.websocket.socketclient.stomp.StompMessage;
import ru.ushell.app.api.websocket.socketclient.stomp.StompMessageListener;

public class TopicHandler {

    private String topic;
    private final Set<StompMessageListener> listeners = new HashSet<>();

    public TopicHandler() {}

    public TopicHandler(String topic) {
        this.topic = topic;
    }

    public String getTopic() {
        return topic;
    }

    public void addListener(StompMessageListener listener) {
        listeners.add(listener);
    }

    public void removeListener(StompMessageListener listener) {
        listeners.remove(listener);
    }

    public void onMessage(StompMessage message){
        for (StompMessageListener listener : listeners) {
            listener.onMessage(message);
        }
    }
}

File: app/src/main/java/ru/ushell/app/data/common/webSocket/TopicHandler.kt
package ru.ushell.app.data.common.webSocket

import ru.ushell.app.data.common.webSocket.stomp.StompMessage
import ru.ushell.app.data.common.webSocket.stomp.StompMessageListener

data class TopicHandler(
    val topic: String,
    val listeners: MutableSet<StompMessageListener> = mutableSetOf()
){
    fun onMessage(message: StompMessage){
        this.listeners.forEach { it.onMessage(message) }
    }

    fun addListener(listener: StompMessageListener){
        this.listeners.add(listener)
    }

    fun removeListener(listener: StompMessageListener){
        this.listeners.remove(listener)
    }
}